pkg/engine/executor/executor.go

```diff
@@ -76,33 +80,124 @@ func (c *Context) executeDataObjScan(ctx context.Context, node *physical.DataObj
+		if streamsSection != nil {
+			return errorPipeline(fmt.Errorf("multiple streams sections found in data object %q", node.Location))
+		}
+
+		var err error
+		streamsSection, err = streams.Open(ctx, sec)
+		if err != nil {
+			return errorPipeline(fmt.Errorf("opening streams section %q: %w", sec.Type, err))
+		}
+	}
+	if streamsSection == nil {
+		return errorPipeline(fmt.Errorf("streams section not found in data object %q", node.Location))
+	}
+
+	for i, sec := range obj.Sections().Filter(logs.CheckSection) {
+		if i != node.Section {
+			continue
+		}
+
+		var err error
+		logsSection, err = logs.Open(ctx, sec)
+		if err != nil {
+			return errorPipeline(fmt.Errorf("opening logs section %q: %w", sec.Type, err))
+		}
+	}
+	if logsSection == nil {
+		return errorPipeline(fmt.Errorf("logs section %d not found in data object %q", node.Section, node.Location))
+	}
+
+	predicates := make([]logs.Predicate, 0, len(node.Predicates))
 
 	for _, p := range node.Predicates {
-		conv, err := buildLogsPredicate(p)
+		conv, err := buildLogsPredicate(p, logsSection.Columns())
 		if err != nil {
 			return errorPipeline(err)
 		}
 		predicates = append(predicates, conv)
 	}
 
-	obj, err := dataobj.FromBucket(ctx, c.bucket, string(node.Location))
-	if err != nil {
-		return errorPipeline(fmt.Errorf("creating data object: %w", err))
-	}
-
-	return newDataobjScanPipeline(dataobjScanOptions{
-		Object:      obj,
+	var pipeline Pipeline = newDataobjScanPipeline(dataobjScanOptions{
+		// TODO(rfratto): passing the streams section means that each DataObjScan
+		// will read the entire streams section (for IDs being loaded), which is
+		// going to be quite a bit of wasted effort.
+		//
+		// Longer term, there should be a dedicated plan node which handles joining
+		// streams and log records based on StreamID, which is shared between all
+		// sections in the same object.
+		StreamsSection: streamsSection,
+
+		LogsSection: logsSection,
 		StreamIDs:   node.StreamIDs,
-		Section:     node.Section,
 		Predicates:  predicates,
 		Projections: node.Projections,
 
-		Direction: node.Direction,
-		Limit:     node.Limit,
+		// TODO(rfratto): pass custom allocator
+		Allocator: memory.DefaultAllocator,
+
+		BatchSize: c.batchSize,
+		CacheSize: 16_000_000, // TODO(rfratto): make configurable at engine level
+	})
+
+	sortType, sortDirection, err := logsSection.PrimarySortOrder()
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "could not determine primary sort order for logs section, forcing topk sort", "err", err)
+	}
+
+	// Wrap our pipeline to enforce expected sorting. We always emit logs in
+	// timestamp-sorted order, so we need to sort if either the section doesn't
+	// match the expected sort order or the requested sort type is not timestamp.
+	//
+	// If it's already sorted, we wrap by LimitPipeline to enforce the limit
+	// given to the node.
+	if node.Direction != logsSortOrder(sortDirection) || sortType != logs.ColumnTypeTimestamp {
```

Comment:
Both this and the else branch need not be applied for metric queries as they apply no Limit and expect to read all matching log lines in any order

Reply:
👍 Thanks, I'll short circuit and return immediately if we detect a metric query.